### PR TITLE
Trimmed-median, trimmed-mean and percentile

### PR DESCRIPTION
**Pull request by ktsaou**
Fixes https://github.com/netdata/netdata/issues/13426

This PR adds the following functionality to the query engine:

- [x] `trimmed-median`, by allowing the `median` function to accept a percentage of the values to be excluded - by giving 5%, the median will be calculated by excluding all points that have a value < min + 5% of delta or > max - 5% of delta, where delta = max - min.
- [x] `trimmed-mean` (by default 5%), which calculates the average excluding the smallest and biggest X% of the points.
- [x] `percentile` (by default 95%), which calculates the average by excluding all the points with the biggest values.

`trimmed-mean` and `percentile` interpolate the last point, when the percentage given does not match exactly the number of points to be included in the calculation.

Also, special care has been given to allow all these algorithms work as expected for both positive and negative dimensions.

It also fixes a couple of bugs in the median algorithm.

### CODE REVIEW

**Comment by ktsaou** [comment]
Ready for review.

File: web/api/queries/trimmed_mean/trimmed_mean.c
Comment (thiagoftsm) [inline web/api/queries/trimmed_mean/trimmed_mean.c:152]: Is there a special reason to use `fprintf` instead `error` or `internal_error`?

**Comment by thiagoftsm** [comment]
Hello @ktsaou,

Code is working as expected, I am only waiting for your answer before to approve it.

File: web/api/queries/trimmed_mean/trimmed_mean.c
Comment (ktsaou) [reply web/api/queries/trimmed_mean/trimmed_mean.c]: It was just left-over debug info. I removed it.

**Comment by ktsaou** [comment]
Fixed. Please review.

**Comment by thiagoftsm** [comment]
> Fixed. Please review.

On it.

**Review by thiagoftsm — APPROVED** [review]
I retest `alert`, dashboard access, and cloud. Everything worked as expected!

**Review by vlvkobal — APPROVED** [review]

File: web/api/netdata-swagger.json
Comment (andrewm4894) [inline web/api/netdata-swagger.json:456]: @hugovalente-pm hey - any idea if, once this merged, we will need to do some netdata cloud BE work to have these new aggregation functions available in here? 

just wondering as im not sure either way. 

image

Comment (hugovalente-pm) [reply web/api/netdata-swagger.json:456]: I don't we will need work on the BE -unless there is some enum limiting the list @netdata/cloud-be to confirm- but I think we will on FE, maybe @burbuli8ra can confirm on FE I think she worked on this

Comment (andrewm4894) [reply web/api/netdata-swagger.json:456]: Cool, I'll see how it goes. That drop dowm might get big now so assuming hopefully it will just have a scrollbar as needed maybe. I'll spin up a node on this branch today to test maybe.

Comment (burbuli8ra) [reply web/api/netdata-swagger.json:456]: Hey @hugovalente-pm !
If I understand correctly we need to add 2 new options on the list shown for time aggregation with the values `percentile` and `trimmed_mean`! If that's the case and those 2 new options are not aliases for already existing values, then yes FE work is needed as well!
We will need to add [here](https://github.com/netdata/charts/blob/f7cdf5e0b597cf78c97063b879e95285a664c041/src/components/line/filterToolbox/timeAggregation.js#L6) the new options and we will also need from you to provide as a `label` (shown inside the dropdown menu) and a `short` label (shown when the dropdown is collapsed and the option selected).

Comment (andrewm4894) [reply web/api/netdata-swagger.json:456]: @burbuli8ra cool good to know. I'll test it a bit on agent end and then follow up with a little PR or issues to have them (or a subset first) available in cloud.

Comment (ktsaou) [reply web/api/netdata-swagger.json:456]: @hugovalente-pm I think we need 3 options:

1. `trimmed-mean 5%`
2. `trimmed-median 5%`
3. `percentile 95%`

It would also be interesting to have a UI way to select other percentiles, at least the predefined ones (1, 2, 3, 5, 10, 15, 20, 25 for the trimmed ones, 25, 50, 75, 80, 90, 95, 97, 98, 99 for percentile).

**Comment by andrewm4894** [comment]
@ktsaou or @thiagoftsm do you know if the functions added by this PR would be available to the health engine? 

Idea being for example to use `percentile` to define an alert that triggers if `$this` value in some alert config crosses the p90 threshold more than x% of time in some window. e.g. one potential candidate default alert for each dim in `apps.cpu` could be if it starts to consistently "cross over" the p90 or p95 threshold for what its used in the past (of course ideally we'd want something symetric such that if is passes either side of p10 or p90 in a persistant manner the alert would trigger - to cover an app dying as well as using too much cpu for example).

I think ideally we need a zscore to do this sort of alerting best but just wanted to check since we already have these percentile functions.

**Pull request merged**
